### ISSUE TITLE
Add a pool for TwistedConnection so that we re-use background threads

### ISSUE THREAD

**Issue by iffy** [alex/alchimia#18]
I keep getting this error when I make many simultaneous queries (klein -> alchimia -> PostgreSQL):

`QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30`

I'm working around it right now by setting `max_overflow=-1` (http://docs.sqlalchemy.org/en/rel_0_8/core/pooling.html#sqlalchemy.pool.QueuePool.params.max_overflow) but I think that will cause problems if I actually open more connections than the Postgres database allows.  So I think I (or someone) could implement a `Deferred`-returning pool that returns `Connections` asynchronously and has requests for connections wait in a queue.

I'm filing this here, because I think alchimia should come with such a pool built in.  I'll take a look at what an implementation might look like on Monday.

**Comment by glyph** [comment]
@iffy It's been a while!  I'm not sure if you're still interested in this.  But there's a new (private, for now) API in Twisted which may be promising for this functionality; `twisted._threads`: https://twistedmatrix.com/documents/16.4.1/api/twisted._threads.html

**Comment by iffy** [comment]
@glyph wow, it has been a while :)  I'm currently not interested (things are working as they are) but if I circle around to needing this again, I'll come back here.

**Comment by glyph** [comment]
@iffy - I actually have an implementation on https://github.com/glyph/alchimia right now, which I hope to integrate and release in the coming week :).  It's untested right now, or rather, it's un-unit-tested, but stress testing with sqlite I can provoke tracebacks right now and with my fork the tracebacks go away.

**Comment by hawkowl** [comment]
@glyph we can close this, right? :)

**Comment by glyph** [comment]
@hawkowl No, because we still haven't dealt with the issue of a `Deferred`-aware pool.  The current strategy, while more correct, is less efficient.  It opens a new thread for every transaction.  Instead, we should recycle idle threads and connections; that's when this should be closed.

**Comment by glyph** [comment]
Yeah, so, it looks like with sqlite it selects the "null pool" http://docs.sqlalchemy.org/en/rel_1_1/dialects/sqlite.html?highlight=sqlite#threading-pooling-behavior - perhaps we should somehow force this to happen with any other database adapters as well, until this can be more thoroughly addressed?
